feat: add frontend support for stock [SCSE-155]

nicolelst opened the pull request:
- added attribute for stock to ProductType, which stores quantity in stock as a 2D array with indices corresponding to (colorway, size)
- on merch details page, 
    - after selecting colorway and size, 
        - display quantity in stock 
        - display quantity in cart (if any)
        - enable purchase quantity input
        - enable add to cart, buy now
    - prevent selecting colorway/size combinations that are out of stock
    - prevent entering purchase quantity beyond (quantity in stock - quantity in cart)
    - disable +/- input buttons if unable to in/decrease purchase quantity 
    - if maximum purchase quantity is already in cart, display warning message and prevent adding to cart
- in cart view,
    - prevent changing purchase quantity beyond quantity in stock
    - display quantity in stock
    - disable +/- input buttons if unable to in/decrease purchase quantity

nicolelst committed:
feat: added mock data for stock

nicolelst committed:
Revert "feat: added mock data for stock"

This reverts commit a81735628c7d9083ce05e7caa4d45f25ddc13f18.

nicolelst committed:
feat: added mock data for stock

Stock for each ProductType is stored as a matrix with indices color,
size.

nicolelst committed:
feat: display stock and disable buttons based on availability

Show remaining quantity based on options selected. Disable size/colorway/quantity input buttons if option is out of stock.

nicolelst committed:
fix: able to deselect sizes/colorways

previously, user is unable to deselect size/colorway. since options are disabled when out of stock, in cases where sizes available for each colorway do not overlap, user is unable to select other colorway even when it is available in other sizes. eg stock = [[1, 2, 0, 0], [0, 0, 3, 4]], since colorway1 and size1 are selected by default, user cannot view sizes available for colorway2.

in this commit, removed default selection of size/colorway, enabled deselection of option, disable option only if entirely OOS, prevent adjustment of qty and add to cart / buy now until all options selected.

nicolelst committed:
fix: max qty affected by qty in cart

consider number already added to cart in calculating maximum purchase quantity

nicolelst committed:
fix: prevent increasing qty in cart beyond qty in stock

nicolelst committed:
feat: display qty in stock in cart

nicolelst commented on src/typings/product.tsx:
```diff
@@ -1,14 +1,17 @@
-  colorways?: string[];
   productCategory?: ProductCategoryType;
-  isAvailable?: boolean;
+  isAvailable: boolean;
```
keeping `isAvailable` as a way to easily suspend/resume the sale of a product. not sure if this is necessary?

chanbakjsd replied:
I'd say we can keep it for now for potential "Coming Soon" items.

nicolelst replied:
In that case, should products that are not available be hidden from the merchandise list (landing page)?

RealDyllon replied:
yup they should be hidden from all views @nicolelst

RealDyllon replied:
may need to hide a product that has been discontinued

nicolelst replied:
Added badge to indicate products are unavailable/OOS from merchandise list page instead

nicolelst replied:
@RealDyllon got it, have hidden unavailable items from merchandise list. 
even if they access the merchandise details page by changing the product id in the url, there should be an `unavailable` badge and the purchasing buttons should be disabled.

nicolelst commented on src/pages/MerchDetail/MerchDetail.tsx:
```diff
@@ -73,25 +75,36 @@ export const MerchDetail: React.FC = () => {
         quantity,
-        size: selectedSize ?? product?.sizes?.[0] ?? "",
-        colorway: selectedColorway ?? ""
+        colorway: selectedColorway ?? "", // TODO ? 
+        size: selectedSize ?? "", // TODO ?
```
@chanbakjsd not sure what the default colorway/size should be
(or if it even matters, since I disabled the "add to cart" and "buy now" buttons until both are selected)

chanbakjsd replied:
You should validate that `selectedColorway` and `selectedSize` is not null, that should allow you to just assign it directly. If it is null, you might want to just return without doing anything.

nicolelst commented on src/pages/MerchDetail/MerchDetail.tsx:
```diff
@@ -73,25 +75,36 @@ export const MerchDetail: React.FC = () => {
+    if (product) {
+      const stockQty = getQtyInStock(product, colorway, size);
+      const cartQty = getQtyInCart(cartState.items, product.id, colorway, size);
+      const max = (stockQty > cartQty) ? stockQty - cartQty : 0;
```
currently setting the maximum quantity to be the quantity in stock - the quantity already added to cart. 
@chanbakjsd should we consider preventing someone from buying the entire stock?
additionally, not sure what happens if 2 people try to do this at the same time.

chanbakjsd replied:
If 2 people attempt to do this at once, the backend should handle this correctly and error as input validation should fail as a result. However, frontend will need to validate which products have ran out when the first person bought all the stock.

nicolelst commented on src/pages/Cart/CartItem.tsx:
```diff
@@ -59,22 +61,29 @@ const CartItem: React.FC<CartItemProps> = ({ isMobile, data, onRemove, onQuantit
+          placeholder="Item Count"
+          onChange={handleQtyChangeInput}
+        />
+        <InputRightAddon style={(data.quantity < MAX_ITEM_CNT) ? { cursor: "pointer" } : { cursor: "not-allowed", opacity: 0.4 }} onClick={() => handleQtyChangeCounter(true)}>
```
decided to grey out the "+" button and change the cursor to "not-allowed" if the user tries to increase the quantity beyond the quantity in stock from the cart view, but we could consider adding a more significant warning/error message? @chanbakjsd

chanbakjsd replied:
I believe this is good enough for now, we can change it in the future if we have a better plan.

chanbakjsd commented on src/pages/MerchDetail/MerchDetail.tsx:
```diff
@@ -155,10 +188,31 @@ export const MerchDetail: React.FC = () => {
+                  false
+                ) ||
+                ((product && selectedSize) ? 
+                  (getQtyInStock(product, colorway, selectedSize) === 0) : // colorway is not available for all sizes
```
This comment seems to be wrong.

chanbakjsd commented on src/data/mock/product/index.ts:
```diff
@@ -10,20 +10,24 @@ export const productList: ProductType[] = [
-    isAvailable: false,
+    price: 2500,
+    colorways: ["Brown"],
+    stock: [[5, 0, 3, 1, 2]],
```
Why are there 5 entries when there are only 4 sizes? In addition, it will also make sense if this entry is merged with `SCSE Sweater - Biege`.

chanbakjsd reviewed (commented):
This is a really good start and I've responded to the concerns you've raised. :)

nicolelst committed:
feat: added unavailable and OOS badges to merchandise list page

when items are unavailable or out of stock, instead of displaying size range on merchandise list, display a badge

nicolelst committed:
fix: add OOS badge to merch detail page & reject add to cart if no selected size/colorway

nicolelst commented:
@chanbakjsd I have made changes according to your feedback, but am not sure what additional validation needs to be implemented in response to 
> frontend will need to validate which products have ran out when the first person bought all the stock.

nicolelst commented on src/pages/MerchandiseList/Card.tsx:
```diff
@@ -36,12 +38,26 @@ const Card = ({ productId, imgSrc, text, price, sizeRange }: CardProps) => {
               <ArrowForwardIcon />
             </Flex>
+            )}
+            <Flex justifyContent="left" mt={1} alignItems="center">
+                {!isAvailable && (
+                  <Badge color="grey" mr={2} fontSize="xs" variant="outline" display="inline">
+                    unavailable
+                  </Badge>
+                )}
+                {isOutOfStock && (
+                  <Badge color="grey" mr={2} variant="outline" display="inline">
+                    out of stock
+                  </Badge>
+                )}
+            </Flex>
```
not sure if having both `unavailable` and `out of stock` badges looks crowded, we can consider prioritising either badge over the other and just displaying one.

nicolelst committed:
Merge branch 'main' into feat/SCSE-155

nicolelst committed:
fix: hide unavailable items from merchandise list

chanbakjsd reviewed (dismissed):
Other than the merge conflict, LGTM!

nicolelst committed:
Merge branch 'main' into feat/SCSE-155
